### ISSUE TITLE
The training time of pretrain models

### ISSUE THREAD

**Issue by leijue222** [s3prl/s3prl#359]
Pretrain upstream models, including Mockingjay, Audio ALBERT and TERA.
**How many** GPUs and **what types** of GPUs did these three models use, and **how long** did they train?

**Comment by leijue222** [comment]
I tried to start the TERA training, adjusted the bs, and used 8*P40(24G) graphics cards. After three days, the training progress was only 3%.
Here is my adjusted config file:
```
runner:
  n_epochs: -1
  total_steps: 1000000
  gradient_clipping: 5.0
  gradient_accumulate_steps: 8

  log_step: 50000
  save_step: 50000
  max_keep: 5

  fp16: false

optimizer:
  name: AdamW_with_schedule
  lr: 2.e-4
  warmup_proportion: 0.07

pretrain_expert:
  datarc:
    num_workers: 8
    train_batch_size: 128
    max_timestep: -200 # Max length for audio feature (0 for no restriction, negative value to set minimum timestep)
    libri_root: 'datasets/LibriSpeech/' # If raw libri data is provided, use on-the-fly feature extraction, else use the pre-extracted features under `file_path`
    file_path: 'data/len_for_bucket' # Pre-extracted features path. When using on-the-fly feature extraction, this is used to provide length for bucketing.
    sets: ['train-clean-100'] # can be the subset of ['train-clean-100', 'train-clean-360', 'train-other-500']
```

**Comment by andi611** [comment]
Hi,

In the TERA paper, we train the model with an actual batch size of 32 (4 * 8) by setting:
```yaml
gradient_accumulate_steps: 4
train_batch_size: 8
```
Please also see here:
https://github.com/s3prl/s3prl/blob/540a4f86e9f099b240c314a56ff80b069e91d4cd/s3prl/pretrain/runner.py#L109-L112
This will take approximately **5 days on a single 1080Ti GPU**.

Note that `gradient_accumulate_steps` are virtual training steps for stacking the batch size when GPU memory is limited. After `gradient_accumulate_steps` amount of virtual steps, the actual training step is increased by one, where the maximum actual training step is defined by `total_steps`.

Looking at your config, your batch size is set to `128*8`=`1024`, which is very large. The training progress at 3% after three days looks reasonable to me at such a large batch size.

To shorten training time, you can either:
1) reduce the batch size, by decreasing `gradient_accumulate_steps` and/or `train_batch_size`
2) reduce `total_steps: 1000000`, since you shouldn't need that many steps for the model to converge if you are training with a large batch size.
3) use more GPUs so you can reduce the `gradient_accumulate_steps`.

Andy

**Comment by leijue222** [comment]
Thank you for your answer, I would like to reproduce your results with the same configuration and as fast as possible.
For `gradient_accumulate_steps`: I would drop it to 1;
For `train_batch_size`: I increased the `bs` to improve GPUS utilization and speed up training. In theory, the larger the bs, the faster the speed. In the case of enough video memory, shouldn't the `bs` be bigger the better?
For `total_steps`: I thought that N GPUs added up to total_steps, so I think the larger the bs, the more GPUs, which will speed up the training. Am I understanding wrong here?

----
(Please ignore the following, it's my machine's problem)
The slowness may also be related to my machine.
Due to my machine, it is very slow to fetch data one by one from the scattered file pile through csv. I seem to need to preprocess all the data into one big file and then index it for training, which will be faster.

**Comment by andi611** [comment]
> the paper setting is 8*1000000, 

I believe it is a batch size of 32 with 1M total steps.

As long as `train_batch_size` times `gradient_accumulate_steps` yields 32 then you are using the same setting as ours, so the following are all correct:
case 1: `train_batch_size=32, gradient_accumulate_steps=1`
case 2: `train_batch_size=16, gradient_accumulate_steps=2`
case 3: `train_batch_size=8, gradient_accumulate_steps=4`
and so on.
In all cases, you should set `total_steps=1000000`.

> If I set `bs=8 -> bs = 256`, then steps should be `8000000/ 256 = 31250`. Rigth?

I understand you want to increase the batch size and thus reduce the training steps,
although the amount of seen samples is the same, please note that the gradients will be different due to the large batch size. (In other words, training the same model with small batch size v.s. large batch size might get very different results.)
That being said, I haven't tried training with a larger batch size so I don't have much to share.

The amount of seen samples of our setting is `32*1000000` and not `8*1000000`.
So if you were to increase batch size from 32 to 256, you can reduce `total_steps` to 1000000/(256/32) = 125000

**Comment by leijue222** [comment]
Thanks,  clearly!
Suppose the GPU has 12G of video memory, but only 2G is used, which is too uncomfortable.
In my past computer vision tasks, the size of the bs is generally adjusted so that the memory usage is close to 11.5/12. 
Usually, a large BS will achieve better results than a small BS, and the training speed is also faster. I don't know if this is the case in voice tasks as well.

I'll update again if it's faster after setup.

**Comment by leijue222** [comment]
Thank you for your careful answer, which made it clearer.
> And FYI, this model can be trained within 5 days on a single GPU if you use our settings.

I'll give it a try. 
By the way, the [TERA default config](https://github.com/s3prl/s3prl/blob/master/s3prl/pretrain/tera/config_runner.yaml) will take approximately 5 days on a single 1080Ti GPU.
How about [mockingjay ](https://github.com/s3prl/s3prl/blob/master/s3prl/pretrain/mockingjay/config_runner.yaml) and [AALBert](https://github.com/s3prl/s3prl/blob/master/s3prl/pretrain/audio_albert/config_runner.yaml)?

**Comment by andi611** [comment]
> Thank you for your careful answer, which made it clearer.
> At this rate, I seem to need 25 days instead of 5.

Ah! I think you are right, I forgot an important factor that effects training time, which is the amount of data. 
5 days of training time is for training on `sets: ['train-clean-100']`, 
25 days of training time is for training on `sets: ['train-clean-100', 'train-clean-360', 'train-other-500']`,
see below for more info, this is related to the `total_steps` it takes for the model to converge for a given amount of data.

> By the way, the [TERA default config](https://github.com/s3prl/s3prl/blob/master/s3prl/pretrain/tera/config_runner.yaml) will take approximately 5 days on a single 1080Ti GPU. How about [mockingjay ](https://github.com/s3prl/s3prl/blob/master/s3prl/pretrain/mockingjay/config_runner.yaml) and [AALBert](https://github.com/s3prl/s3prl/blob/master/s3prl/pretrain/audio_albert/config_runner.yaml)? 

They are almost the same, since they are all 3-layer transformer encoders, and training objectives are similar.

> In addition, now I only used the `train-clean-100`, the `train-clean-360` and `train-Other-500` I did not join them. The paper should have used 960h, but I only used 100h now. (At first, I just want to be faster. I will increase it to 960h to keep the same after confirming that the configuration and the machine are ok.)

If you use `sets: ['train-clean-100', 'train-clean-360', 'train-other-500']`, 
then use `total_steps: 1000000` for the model to converge, which is approximately 25 days.
If `sets: [train-clean-360']`, then `total_steps: 500000` is enough (empirically).
If `sets: [train-clean-100']`, then `total_steps: 200000` is enough (empirically), which is approximately 5 days.

**Comment by leijue222** [comment]
200000

**Comment by leijue222** [comment]
I got it!
My config on TERA now is 
```
TERA
single GPU, 16/24G
total_steps: 1000000,gradient_accumulate_steps: 4, train_batch_size: 8, sets: ['train-clean-100']
{"@timestamp":"2022-06-16 10:02:51.304","@message":"\roverall:   0%|          | 0/1000000 [00:00<?, ?it/s]"}
{"@timestamp":"2022-06-17 09:31:52.716","@message":"\rtrain:  70%|███████   | 2495/3564 [20:47<09:14,  1.93it/s]\u001b[A\roverall:   4%|▍         | 42501/1000000 [23:29:01<532:30:39,  2.00s/it]"}
```
23h31min=1411min 42501 steps. At this rate, I seem to need 25 days. I should set `total_steps: 200000` if I only use `sets: ['train-clean-100']` on a single GPU.
**By the way, based on your experience, how to set `gradient_accumulate_steps=?` and `train_batch_size=?` when using 4GPU or 8GPU or more on pretain model trainning?**

----

Moreover, I also test on sv task based on wavlm_large.
`python run_downstream.py -m train -u wavlm_large -d sv_voxceleb1 -n wavlm_large_single`
`python -m torch.distributed.launch --nproc_per_node 4 run_downstream.py -m train -n wavlm_large_ddp4 -u wavlm_large -d sv_voxceleb1 \
    -o config.runner.gradient_accumulate_steps=2`
```
Wavlm_large for sv_voxceleb1
single GPU， 4.5/24G
config default: gradient_accumulate_steps: 5, train_batch_size: 10
{"@timestamp":"2022-06-17 00:13:25.409","@message":"\roverall:   0%|          | 0/200000 [00:00<?, ?it/s]"}
{"@timestamp":"2022-06-17 09:30:20.557","@message":"\rtrain:  15%|█▍        | 2262/15349 [26:04<2:30:33,  1.45it/s]\u001b[A\roverall:   5%|▍         | 9662/200000 [9:16:55<182:44:30,  3.46s/it]"}
9h17min=557min 9662 steps

----
Wavlm_large for sv_voxceleb1
single GPU， 22/24G
config change: gradient_accumulate_steps: 2, train_batch_size: 80
{"@timestamp":"2022-06-16 19:38:26.416","@message":"\roverall:   0%|          | 0/200000 [00:00<?, ?it/s]"}
{"@timestamp":"2022-06-17 09:31:22.989","@message":"\rtrain:  60%|█████▉    | 1145/1919 [1:48:14<1:13:07,  5.67s/it]\u001b[A\roverall:   2%|▏         | 4411/200000 [13:52:56<618:46:48, 11.39s/it]"}
13h43min=823min 4411 steps

----
Wavlm_large for sv_voxceleb1
4 GPU DDP, each 22/24G
config change: gradient_accumulate_steps: 2, train_batch_size: 80
{"@timestamp":"2022-06-16 19:42:41.434","@message":"\roverall:   0%|          | 0/200000 [00:00<?, ?it/s]"}
{"@timestamp":"2022-06-17 09:28:43.286","@message":"\rtrain:   1%|          | 3/480 [00:24<58:37,  7.37s/it]  \u001b[A\roverall:   2%|▏         | 4322/200000 [13:46:02<694:23:39, 12.78s/it]"}
13h46min=826min 4322 steps
```
We can see that:
1. compare single GPU case, one is default `gradient_accumulate_steps: 5, train_batch_size: 10`,  it reached 9662 steps in less time; The other one is `gradient_accumulate_steps: 2, train_batch_size: 80`. It only reached 4411 steps and took more time. In here, The larger the bs, the slower the training speed.
2. conpare single GPU with 4 GPU DDP, they are same `gradient_accumulate_steps: 2, train_batch_size: 80` and the also take nearly the same number of steps in the same amount of time. Which means under this config that there is no speed increase for multi-GPU training.

Conclusions:

- Turning up the BS to fill up the video memory GPU does not play a role in speeding up;
- Using the default configuration of s3prl, although it only takes up a small part of the GPU memory, it is faster.

I suddenly remembered that there is Batch Normalization in visual tasks. The larger the BS, the better the features can be learned and the overfitting can be avoided.
Speech is a sequence model. Due to the different lengths, there is no Batch Normalization in the sequence model. The larger the BS, the less frequently the parameters are updated, and the lack of BN for normalization will make the model more difficult to learn. Therefore, the utilization of GPU memory should not be pursued blindly. After all, the larger the BS, the slower the speed, and the effect may be worse.

Again, based on your experience, how to set gradient_accumulate_steps=? and train_batch_size=? when using 4GPU or 8GPU or more on this ASV task?

**Comment by andi611** [comment]
> We can see that:
> 
> 1. compare single GPU case, one is default `gradient_accumulate_steps: 5, train_batch_size: 10`,  it reached 9662 steps in less time; The other one is `gradient_accumulate_steps: 2, train_batch_size: 80`. It only reached 4411 steps and took more time. In here, The larger the bs, the slower the training speed.
> 2. compare single GPU with 4 GPU DDP, they are same `gradient_accumulate_steps: 2, train_batch_size: 80` and the also take nearly the same number of steps in the same amount of time. Which means under this config that there is no speed increase for multi-GPU training.
> 
> Conclusions:
> 
> * Turning up the BS to fill up the video memory GPU does not play a role in speeding up;
> * Using the default configuration of s3prl, although it only takes up a small part of the GPU memory, it is faster.

@leo19941227 Can you comment on this? 
Why is `gradient_accumulate_steps: 2, train_batch_size: 80` slower than `gradient_accumulate_steps: 5, train_batch_size: 10`?
Is it the communication cost between devices that are slowing the process down? 
Or maybe the cost to extract features on-the-fly for a large batch size is too high?
I believe you are more experienced with DDP than I am. 

> I suddenly remembered that there is Batch Normalization in visual tasks. The larger the BS, the better the features can be learned and the overfitting can be avoided. Speech is a sequence model. Due to the different lengths, there is no Batch Normalization in the sequence model. The larger the BS, the less frequently the parameters are updated, and the lack of BN for normalization will make the model more difficult to learn. Therefore, the utilization of GPU memory should not be pursued blindly. After all, the larger the BS, the slower the speed, and the effect may be worse.

I believe we can also do Batch Normalization on sequential data.

> **Again, based on your experience, how to set `gradient_accumulate_steps=?` and `train_batch_size=?` when using 4GPU or 8GPU or more on this ASV task?** (make training faster)

Maybe Leo could comment on this one as well!

**Comment by leo19941227** [comment]
Since it is important to consider the I/O and GPU utility balance when training your models.
In the first config, you must first load 80 data points before feed them into GPU, hence the GPU waits so long (utility 0) before all the 80 data points are loaded.
In the second config, you feed the data points for GPU as soon as 10 samples are loaded, during the GPU forward, you continue to load more data points, hence now both GPU and I/O have proper utility, hence it is faster.
In the first config, you are wasting GPU power most of the time due to I/O bottleneck.
The first config can ideally faster than the second config when all the system resources (I/O, GPU, CPU) can get 100% utility all the time, however this put huge pressure on disk I/O speed and is only possible when you are using SSD.

**Comment by leijue222** [comment]
Hi @andi611, I followed your advice and trained TERA with the default configuration.
On [SUPERB](https://superbbenchmark.org/leaderboard): TERA for SV the ERR is **15.89**.
I use officeal tera_100hr to finetune downstream task of SV: `python run_downstream.py -m train -u tera_100hr -d sv_voxceleb1 -n tera_100hr_G1_5_10`, the ERR is **29.83**. The result is slower than on SUPERB.  
**I guess the reason is that I used tera_100hr as base model to finetune SV rather use tera as base model.** 
Then I train TERA from scratch with the default configuration(only change sets to train_clean_100h), using TERA trained for 350K steps as the base model finetune SV. The ERR is **9.45**, The result is higher than on SUPERB.
**I guess offical tera_100hr model only trained 200K steps, but I train 350K steps.**
Is my understanding correct?
